src/sources/kubernetes.rs

```diff
@@ -0,0 +1,318 @@
+struct PodData {
+    /// This exists because Docker is still a special entity in Kubernetes as it can write in Json
+    /// despite CRI defining it's own format.
+    log_format: Option<LogFormat>,
+}
+
+struct TimeFilter {
+    start: DateTime<Utc>,
+}
+
+impl TimeFilter {
+    fn new() -> Self {
+        // Only logs created at, or after this moment are logged.
+        let now = Utc::now();
+        TimeFilter { start: now }
+    }
+
+    fn filter(&self, event: Event) -> Option<Event> {
+        // Only logs created at, or after now are logged.
+        if let Some(ValueKind::Timestamp(ts)) = event.as_log().get(&event::TIMESTAMP) {
+            if ts >= &self.start {
+                return Some(event);
+            }
+            trace!(message = "Recieved older log", from = %ts.to_rfc3339());
+        }
+        None
+    }
+}
+
+fn file_source(
+    kube_name: &str,
+    globals: &GlobalOptions,
+) -> Result<(mpsc::Receiver<Event>, super::Source), String> {
+    let mut config = FileConfig::default();
+
+    // TODO: Having a configurable option for excluding namespaces, seams to be usefull.
+    // // TODO: Find out if there are some guarantee from Kubernetes that current build of
+    // // TODO  pod_uid as namespace_pod-name_some-number is a somewhat lasting decision.
+    // NOTE: pod_uid is unspecified and it has been found that on EKS it has different scheme.
+    // NOTE: as such excluding/including using path is hacky, instead, more proper source of
+    // NOTE: information should be used.
+    // NOTE: At best, excluding/including using path can be an optimization
+    // TODO: Exclude whole kube-system namespace properly
+    config
+        .exclude
+        .push((LOG_DIRECTORY.to_owned() + r"kube-system_*/**").into());
+    // TODO: Add exclude_namspace option, and with it in config exclude namespace used by vector.
+    // NOTE: for now exclude images with name vector, it's a rough solution, but necessary for now
+    config
+        .exclude
+        .push((LOG_DIRECTORY.to_owned() + r"*/vector*/*").into());
+
+    config
+        .include
+        .push((LOG_DIRECTORY.to_owned() + r"*/*/*.log").into());
+
+    config.start_at_beginning = true;
+
+    // Filter out files that certainly don't have logs newer than now timestamp.
+    config.ignore_older = Some(10);
+
+    // oldest_first false, having all pods equaly serviced is of greater importance
+    //                     than having time order guarantee.
+
+    // CRI standard ensures unique naming.
+    config.fingerprinting = FingerprintingConfig::DevInode;
+
+    // Have a subdirectory for this source to avoid collision of naming its file source.
+    config.data_dir = Some(globals.resolve_and_make_data_subdir(None, kube_name)?);
+
+    let (file_send, file_recv) = mpsc::channel(1000);
+    let file_source = config
+        .build("file_source", globals, file_send)
+        .map_err(|e| format!("Failed in creating file source with error: {:?}", e))?;
```

Comment:
we should start using the errors that #811 brought.

Reply:
Done.

Also, this is a good use case for an error with a source error, but I don't see it used in Vector.